### PR TITLE
Background rendering

### PR DESCRIPTION
**Pull request by iBicha**
This PR introduces a couple of fixes:

- Removing event handlers from the frame and plane events when stopping the service
- Implementing `TryGetCameraImage` on the remote editor, which returns the last frame sent from the device.

But mainly, this PR refactors the background rendering.
Replaced all background renderings with the `UnityEngine.XR.ARBackgroundRenderer` which allowed a consistent way to render backgrounds, as opposed to using different scripts on different platforms. (They are different, but they are all using command buffers, which is why using the `ARBackgroundRenderer` was perfect for this use case)

In addition, this allowed me to easily add a new property to the interface: `bool BackgroundRendering` which returns false by default, and is up to the interface to implement it.
This option is now also exposed in the `ARController` and therefore the `ARRemoteEditor`

Toggling the background rendering is as easy as calling:
```
public void Toggle () {
    UnityARInterface.ARInterface.GetInterface().BackgroundRendering = !UnityARInterface.ARInterface.GetInterface().BackgroundRendering;
}
```

The motivation behind this is how the rendering was limiting me from getting creative with the interface. Here's a small example of why developers might need to turn off background rendering: (turn up your volume!)

[IMAGE ALT TEXT HERE](https://www.youtube.com/watch?v=HI-a4pMorqE)

Tested on:

- [x] ARCore
- [ ] ARKit (should work, since it uses exactly the same code and material as the remote)
- [x] Remote

### CODE REVIEW

File: Assets/UnityARInterface/ARRemote/Scripts/ARRemoteVideo.cs [hunk 1]
```diff
@@ -1,58 +0,0 @@
-using System;
```
Comment (jimmy-jam) [inline Assets/UnityARInterface/ARRemote/Scripts/ARRemoteVideo.cs]: why was this file removed?  was it not being used?

Comment (iBicha) [reply Assets/UnityARInterface/ARRemote/Scripts/ARRemoteVideo.cs]: It was, but not anymore. After refactoring the background rendering, this file is no longer being used, because all the interfaces are now rendering background using `UnityEngine.XR.ARBackgroundRenderer`.

**Comment by cgfarmer4** [comment]
@iBicha so the only fallout I see in ARKit since updating is the YUVMaterial flashes on the screen before rendering begins. 

This did not happen before and is not a desired effect. I don't see this in ARCore. 

My best guess is it seems to be a race condition where the material is being setup before the camera? Any suggestions?

**Comment by iBicha** [comment]
@cgfarmer4 thank you for the feedback! For how long the "flashing" goes on?

**Comment by cgfarmer4** [comment]
@iBicha Extremely brief (~1s) but enough to jolt the UX.

**Comment by shawmakesmusic** [comment]
@iBicha Have been using the ARInterface with the new Lightweight SRP. Everything works except for the background rendering. Does this fix that? Or is there a fix? :)

**Comment by cgfarmer4** [comment]
@iBicha the issue is that when you removed the `UnityARVideo` class from `SetupCamera` and ported the `OnPreRender` function, you omitted the CommandBuffer. As best I can deduce, the CommandBuffer tells the renderer when its ready (hence no green screen).

Previous setup camera::

https://github.com/Unity-Technologies/experimental-ARInterface/blob/e7b3dd2e9437c79e4725b6ae9180adeec09e7902/Assets/UnityARInterface/Scripts/ARKitInterface.cs#L207

Unity ARVideo function::

https://bitbucket.org/Unity-Technologies/unity-arkit-plugin/src/59dd4d0ecfc67101378c126aff175775d506ef9c/Assets/UnityARKitPlugin/Plugins/iOS/UnityARKit/Helpers/UnityARVideo.cs?at=default&fileviewer=file-view-default#UnityARVideo.cs-61

```
if (!bCommandBufferInitialized) {
    InitializeCommandBuffer ();
 }
```

You likely omitted this since ARCore does not use that `UnityARVideo` class but that command is important for iOS rendering.

I'm not sure how to best resolve with the new background rendering class. Can you suggest a fix please as this is inhibiting my ability to use the latest ARInterace with ARCore1.0?

**Comment by iBicha** [comment]
@shawmakesmusic I didn't get the chance to try this with the new SRP, but I'm willing to, as soon as I get some free time

@cgfarmer4 With this PR I did not exclude the CommandBuffer, in fact, background rendering does not work at all without it.

The way it works, we tell the camera to do one thing before it starts rendering the scene:

```
m_VideoCommandBuffer = new CommandBuffer(); 
m_VideoCommandBuffer.Blit(null, BuiltinRenderTextureType.CurrentActive, m_ClearMaterial);
GetComponent<Camera>().AddCommandBuffer(CameraEvent.BeforeForwardOpaque, m_VideoCommandBuffer);
```
Which is "Hey, before you start rendering opaque objects, do blit command to fill the screen using the `m_ClearMaterial` material, in which the camera texture is set.

The other interesting this we can do to optimize, is to disable rendering the skybox, since we are rendering the camera texture on top of it, which is a waste. For that purpose, we set the camera clear flag to depth only.

Now this is the main functionality of the `UnityARVideo` component, with one addition, which is assigning the camera texture to the material of the command buffer, so if the camera updates the texture, these changes get rendered.

The thing is, this is EXACTLY what the [UnityEngine.XR.ARBackgroundRenderer](https://docs.unity3d.com/ScriptReference/XR.ARBackgroundRenderer.html) class does, and it does it really well. It works with both forward and deferred rendering, and caches and restores the camera clear flag, has an event to notify you when background rendering changes (if you need that) and it's as easy as assigning a material and a camera to it, and then assigning the rendering [mode](https://docs.unity3d.com/ScriptReference/XR.ARBackgroundRenderer-mode.html) to either [StandardBackground](https://docs.unity3d.com/ScriptReference/XR.ARRenderMode.StandardBackground.html) or [MaterialAsBackground](https://docs.unity3d.com/ScriptReference/XR.ARRenderMode.MaterialAsBackground.html). It really  does all the work for us, so we don't need to mess with any command buffers. ARCore is actually using it since the first version, so why not leverage it on all platforms, so we would have clean, refactored code, and the ability to switch on and off background rendering consistently across platforms (and we don't have to maintain code that is already built-in within Unity).

So, that's for explaining what this PR is about, I did not had time to investigate the bug that you mentioned, but I do suspect something, and it would be the first thing I would check out; On the ARKit session setup, the interface allocates memory to hold the texture data. And right after, this memory gets loaded into a texture, and loaded in the material of the background renderer. I think this happens before the first frame of the camera is available (before the first call to `UpdateFrame`, which initialises and updates the textures), and that makes it render what's in memory, and not valid image. 
The old `UnityARVideo` component only initialises command buffers **after** the camera textures are ready (On its `OnPreRender` function, it literally does nothing if `textureY` or `textureCbCr` are null).

**Comment by Ursanon** [comment]
@cgfarmer4  I made PR with (temporary?)hotfix for iOS green screen problem (#46).

@iBicha It seems that we have nullptr in first frame. That's why we see green screen - need more work at this.

**Comment by cgfarmer4** [comment]
@iBicha thanks so much for the long explanation. I am in my first year of graphics programming so I really appreciate going the extra step to make it clear! I'll take a look @Ursanon.
